File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/video/QualitySelector.kt
package dev.hebei.camerax_android.video

import android.util.Size
import dev.hebei.camerax_android.core.CameraInfo

class QualitySelector internal constructor(internal val obj: androidx.camera.video.QualitySelector) {
    companion object {
        fun from(quality: Quality): QualitySelector {
            val obj = androidx.camera.video.QualitySelector.from(quality.obj)
            return QualitySelector(obj)
        }

        fun from(quality: Quality, fallbackStrategy: FallbackStrategy): QualitySelector {
            val obj = androidx.camera.video.QualitySelector.from(quality.obj, fallbackStrategy.obj)
            return QualitySelector(obj)
        }

        fun fromOrderedList(qualities: MutableList<Quality>): QualitySelector {
            val obj =
                androidx.camera.video.QualitySelector.fromOrderedList(qualities.map { it.obj })
            return QualitySelector(obj)
        }

        fun fromOrderedList(
            qualities: MutableList<Quality>, fallbackStrategy: FallbackStrategy
        ): QualitySelector {
            val obj = androidx.camera.video.QualitySelector.fromOrderedList(
                qualities.map { it.obj }, fallbackStrategy.obj
            )
            return QualitySelector(obj)
        }

        fun getResolution(cameraInfo: CameraInfo, quality: Quality): Size? {
            return androidx.camera.video.QualitySelector.getResolution(cameraInfo.obj, quality.obj)
        }
    }
}

File: camerax/lib/src/video/quality_selector.dart
import 'package:camerax/src/core.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart'
    as $interface;
import 'package:meta/meta.dart';

import 'fallback_strategy.dart';

final class QualitySelector {
  final $interface.QualitySelector _obj;

  QualitySelector._native(this._obj);

  QualitySelector.from(
    $interface.Quality quality, {
    FallbackStrategy? fallbackStrategy,
  }) : _obj = $interface.QualitySelector.from(
          quality,
          fallbackStrategy: fallbackStrategy?.obj,
        );

  QualitySelector.fromOrderedList(
    List<$interface.Quality> qualities, {
    FallbackStrategy? fallbackStrategy,
  }) : _obj = $interface.QualitySelector.fromOrderedList(
          qualities,
          fallbackStrategy: fallbackStrategy?.obj,
        );

  static Future<$interface.Size?> getResolution(
          CameraInfo cameraInfo, $interface.Quality quality) =>
      $interface.QualitySelector.getResolution(cameraInfo.obj, quality);

  @internal
  $interface.QualitySelector get obj => _obj;
}

extension QualitySelectorObj on $interface.QualitySelector {
  @internal
  QualitySelector get args {
    return QualitySelector._native(this);
  }
}
